src/core/main.c

```diff
@@ -1253,37 +1253,39 @@ static void bump_file_max_and_nr_open(void) {
 #endif
 
 #if BUMP_PROC_SYS_FS_NR_OPEN
-        int v = INT_MAX;
-
-        /* Argh! The kernel enforces maximum and minimum values on the fs.nr_open, but we don't really know
-         * what they are. The expression by which the maximum is determined is dependent on the architecture,
-         * and is something we don't really want to copy to userspace, as it is dependent on implementation
-         * details of the kernel. Since the kernel doesn't expose the maximum value to us, we can only try
-         * and hope. Hence, let's start with INT_MAX, and then keep halving the value until we find one that
-         * works. Ugly? Yes, absolutely, but kernel APIs are kernel APIs, so what do can we do... 🤯 */
+        /* cf. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/file.c?h=v6.8#n27
+	 * for the progeny of the below value for `sysctl_nr_open_max`. Note
+	 * that the below logic was first introduced in `git` commit
+	 * eceea0b3df05ed262ae32e0c6340cc7a3626632d of the `linux` kernel and
+	 * was designed to prevent overflows when determining the maximum
+	 * number of possible file descriptors. */
+	#define BITS_PER_LONG __WORDSIZE
+        #define __const_min(x, y) ((x) < (y) ? (x) : (y))
+        unsigned sysctl_nr_open_max =
+            __const_min(INT_MAX, ~(size_t)0/sizeof(void *)) & -BITS_PER_LONG;
```

Comment:
This is not so complex, please rewrite to make it follow our coding style and comment the original code from the kernel. How about something like the following?
```c
/* From fs/file.c in the kernel,
 * unsigned sysctl_nr_open_max = __const_min(INT_MAX, ~(size_t)0/sizeof(void *)) & -BITS_PER_LONG; */
int v = CONST_MIN((size_t) INT_MAX, SIZE_MAX / sizeof(void*)) & ~(sizeof(long) * 8 - 1)
```
Please use `int`, rather than `unsigned`. It is limited to INT_MAX.

Reply:
With all due respect, one of the concerns in adopting this code is that it runs the risk of diverging from the kernel source. So, as long as we're violating the soft guideline/principal which seeks to avoid vendoring expressions from foreign codebases, I'd prefer to keep it as consistent as possible against the kernel source.

If this change set will not be adopted unless I refuse to coerce the `linux` expression which has stood for almost 20 years into a shape that's compatible with `systemd`'s opinionated style guide, then I'll coerce this expression (running the risk of having this break in unexpected ways which I'm not prepared to anticipate considering I can't confirm that your expression is equivalent to `linux`'s for all architectures and all compilers). But, I'm strongly opposed to this change, in principal, so I'd only do this under duress.

Reply:
Please don't blindly copy, but respect our coding style. The suggested change is also way more readable than redefining all those macros and such.

There's no need to keep things in sync bit by bit.

Reply:
> Please don't blindly copy.
I don't know what this means. But, if it means: "copy without consideration of what exactly it is that you're copying" then we're good - I didn't do that. I intentionally used exactly the same definitions from the kernel for consistency.

> respect our coding style

I have been trying. That's why 4673cac7612b0dbaaca6fc78cf03586395b1d810, b4ed665390d1a752d76a081d2fe867e4376ed2fe a989acf11b03a88b3eb0e4120d9cf352c4f4bd21 and others exist, now.

> The suggested change is also way more readable than redefining all those macros and such.

That may be the case. But readability isn't my goal - consistency with the kernel source is my goal.

>

Reply:
They are equal (except for `~(sizeof(long) * 8 - 1)` -> `-(sizeof(long) * 8 - 1)`, should be a typo)

> I'm hesitant to even consider this change since the type that I'm being asked to use (int) is even different than the kernel's (unsigned int). 

It won't make any difference, since the value is bound to be smaller than `INT_MAX` (wrt `CONST_MIN`)

Reply:
> They are equal (except for `~(sizeof(long) * 8 - 1)` -> `-(sizeof(long) * 8 - 1)`, should be a typo)

I intentionally changed that, and these should be effectively equivalent. I just want to not use negative value for unsigned type.